### PR TITLE
feat(drivers): fetch secrets from google secret manager

### PR DESCRIPTION
**Pull request by Charles546**
#### Description

The new driver should provide the capability to fetch secrets from
google secret manager.

The daemon code is updated to support passing non-base64 string to the
driver for lookup instead of decryption.

#### This PR fixes the following issues
N/A

### CODE REVIEW

File: internal/service/service.go [hunk 1]
```diff
@@ -150,10 +150,16 @@ func (s *Service) decryptDriverData(key string, val interface{}) (ret interface{
...
-		dipper.Logger.Panicf("encrypted data shoud be base64 encoded")
+	var decoded []byte
+
+	if parts[1][0] == '"' {
+		decoded = []byte(strings.Trim(parts[1], "\""))
+	} else {
+		var err error
+		decoded, err = base64.StdEncoding.DecodeString(parts[1])
+		if err != nil {
+			dipper.Logger.Panicf("encrypted data shoud be base64 encoded")
```
Comment (wyardley) [inline internal/service/service.go:161]: minor typo (shoud vs should) (exists in the source and the other place this error exists as well)

File: internal/service/service.go [hunk 2]
```diff
@@ -150,10 +150,16 @@ func (s *Service) decryptDriverData(key string, val interface{}) (ret interface{
...
 	}
 
-	data := []byte(parts[1])
-	decoded, err := base64.StdEncoding.DecodeString(string(data))
-	if err != nil {
-		dipper.Logger.Panicf("encrypted data shoud be base64 encoded")
+	var decoded []byte
+
+	if parts[1][0] == '"' {
+		decoded = []byte(strings.Trim(parts[1], "\""))
```
Comment (wyardley) [inline internal/service/service.go:156]: can this use `'"'` as well for consistency with the `if` condition (or vice-versa)?

This whole part seems a little obscure - maybe you can add a comment explaining what's happening here and what `parts[1][0]` is expected to be, or (better), make the whole thing less magic? Is this the case when we have a string starting with a leading double-quote instead of a base64 encrypted string?

File: pkg/dipper/encryption.go [hunk 3]
```diff
@@ -26,10 +26,16 @@ func DecryptAll(rpc RPCCaller, from interface{}) {
...
+		var decoded []byte
+
+		if parts[1][0] == '"' {
+			decoded = []byte(strings.Trim(parts[1], "\""))
+		} else {
+			var err error
+			decoded, err = base64.StdEncoding.DecodeString(parts[1])
+			if err != nil {
+				Logger.Panicf("encrypted data shoud be base64 encoded")
+			}
```
Comment (wyardley) [inline pkg/dipper/encryption.go:38]: this whole routine seems exactly the same as the one in `internal/service`? should it be a common function?

**Review by wyardley — COMMENTED** [review]
Presumably there are some docs changes / examples needed here? And would `gcloud-secret-manager` be more consistent than `gcloud-secret`?

high level, is the interface like`ENC[gcloud-secret-manager,secretname]`? I wonder if a reference to a secret should be treated differently from an encrypted string entirely; in other words, maybe we should have a different way of specifying a secret manager secret without using `[ENC]` at all, since we're not including an encrypted string at all...
In other words, maybe we should have a different way of specifying a secret manager secret entirely.

Either way, can we have a test case for whatever method is handling the `ENC[xyz]` strings that covers these different cases to at least make sure we're grabbing the correct parts, and maybe some cases with bogus data as well? This would also help to make it clearer what the code is doing and what the input is expected to look like.

File: internal/service/service.go (hunk 2 above)
Comment (Charles546) [reply internal/service/service.go:156]: Right, if it starts with a double quote, then I don't use base64 to decode it.

File: pkg/dipper/encryption.go (hunk 3 above)
Comment (Charles546) [reply pkg/dipper/encryption.go:38]: Merged them.

File: internal/service/service.go (hunk 1 above)
Comment (Charles546) [reply internal/service/service.go]: Fixed.

**Comment by Charles546** [comment]
> Presumably there are some docs changes / examples needed here? And would `gcloud-secret-manager` be more consistent than `gcloud-secret`?
The `manager` just doesn't add any value here. The Secret Manager calls the items `secrets` in their resource IDs anyway.

> 
> high level, is the interface like`ENC[gcloud-secret-manager,secretname]`? I wonder if a reference to a secret should be treated differently from an encrypted string entirely; in other words, maybe we should have a different way of specifying a secret manager secret without using `[ENC]` at all, since we're not including an encrypted string at all...
> In other words, maybe we should have a different way of specifying a secret manager secret entirely.
>
Didn't we learn the `ENC[...]` eyaml notion from Datadog? Didn't we also use this to wrap a name in the Datadog helper? I can use a `(str)` prefix, like in `ENC[gcloud-secret,(str)projects/...]`, instead of double quotes, if you think that makes it less magic.
 
> Either way, can we have a test case for whatever method is handling the `ENC[xyz]` strings that covers these different cases to at least make sure we're grabbing the correct parts, and maybe some cases with bogus data as well? This would also help to make it clearer what the code is doing and what the input is expected to look like.
I can add a test for that.

**Comment by wyardley** [comment]
> Didn't we learn the ENC[...] eyaml notion from Datadog? Didn't we also use this to wrap a name in the Datadog helper? I can use a (str) prefix, like in ENC[gcloud-secret,(str)projects/...], instead of double quotes, if you think that makes it less magic.

The concept is based more on Puppet's implementation in hiera 

https://github.com/voxpupuli/hiera-eyaml

The first part has more to do with the specific cipher, but it's always followed with ciphertext, IIRC. To me, getting a value from secret manager is a different concept, so it might be better to have a different way of specifying it (some kind of special value like a secretref).

**Comment by Charles546** [comment]
How about this?

`LOOKUP[google-secret,projects/foo/secrets/bar/versions/latest]`

**Comment by wyardley** [comment]
> LOOKUP[google-secret,projects/foo/secrets/bar/versions/latest]

That seems like a good way to do it.

**Comment by Charles546** [comment]
Cool, I will work on some tests.

**Comment by Charles546** [comment]
Test added.

File: pkg/dipper/encryption_test.go [hunk 4]
```diff
@@ -109,3 +109,49 @@ data:
 	assert.Equal(t, "ENC[driver1,YWFiYmNjZGQ=]", MustGetMapDataStr(data, "data.item3.item4"), "item4 should be stripped off one deferred flag")
 	assert.Equal(t, "not encrypted", MustGetMapDataStr(data, "data.item1"), "data.item1 should remain unchanged")
 }
+
+func TestDecryptAllWithLookUp(t *testing.T) {
+	doc := `
+data:
+  item1: not encrypted
+  item2: LOOKUP[kvstore,foo]
```
Comment (wyardley) [inline pkg/dipper/encryption_test.go:117]: can we test something with the `ENC` prefix too? and testing around base64 encoded (or not) strings?

Comment (Charles546) [reply pkg/dipper/encryption_test.go:117]: There have been tests for `ENC[` in the same file from before.

**Review by wyardley — APPROVED** [review]
